forest/colors.py

```diff
@@ -156,35 +155,36 @@ def reducer(state, action):
     return state
 
 
-@middleware
-def palettes(store, next_dispatch, action):
+def palettes(store, action):
     """Color palette middleware
 
     Encapsulates colorbar user interface logic. For example,
     if a user has chosen to fix their data limits, then
     set_limit actions generated by column data source changes
     are ignored
+
+    .. note:: middleware is an action generator
     """
     kind = action["kind"]
     if kind == SET_PALETTE:
         payload = action["payload"]
         if "name" in payload:
             name = payload["name"]
             numbers = palette_numbers(name)
-            next_dispatch(set_palette_numbers(numbers))
+            yield set_palette_numbers(numbers)
             if "colorbar" in store.state:
                 if "number" in store.state["colorbar"]:
                     number = store.state["colorbar"]["number"]
                     if number not in numbers:
-                        next_dispatch(set_palette_number(max(numbers)))
-        next_dispatch(action)
+                        yield set_palette_number(max(numbers))
+        yield action
     elif kind == SET_LIMITS:
         if is_fixed(store.state) and is_source_origin(action):
             # Filter SET_LIMIT actions from ColumnDataSource
-            return
-        next_dispatch(action)
+            return []  # Syntax to return null generator
```

Comment:
1. Agreed I only remembered post commit that all `return` does inside a generator is raise a `StopIteration`
2. What is your view on guard clauses? Used appropriately they can improve grokability, e.g. exit early don't make me read more of the code or to simplify nested if statements. But admittedly I have over-used them in more than one place in the code base since they are hindering readability

Reply:
Personally I try to avoid multiple returns:
1. It's not uncommon to want to do something common with all the return values of a function, even if it's just printing them out during development.
2. This is obviously subjective, but I usually find it easier to see at a glance what a single-exit routine is doing, i.e. it comes in the top, does stuff, and exits at the bottom. But that readability is also dependent on keeping the complexity of the control flow down, e.g. by factoring it out into subroutines.

Reply:
I hoisted the guard clause logic up to the top of the routine to make it really clear to the reader that if we receive a certain criterion we filter the action.